src/Microsoft.OData.Client/Serialization/Serializer.cs

```diff
@@ -358,12 +358,162 @@ internal void WriteEntry(EntityDescriptor entityDescriptor, IEnumerable<LinkDesc
+                    }
+                    else if (item.DescriptorKind == DescriptorKind.Link)
+                    {
+                        odataWriter.WriteEntityReferenceLink(new ODataEntityReferenceLink() { Url = navigationLink.Url }, relatedEnd.Source, relatedEnd.Target);
+                    }
+                }
+
+                odataWriter.WriteNestedResourceInfoEnd(navigationLink, relatedEnd.Source, relatedEnd.Target);
+                odataWriter.WriteNestedResourceInfoEnd();
+            }
+        }
+
         /// <summary>
         /// Writes the feed entry
         /// </summary>
-        /// <param name="descriptors">A list of descriptors to be written</param>
-        /// <param name="linkDescriptors">Descriptor links for some descriptors.</param>
-        /// /// <param name="bulkUpdateGraph">An instance of the <see cref="BulkUpdateGraph"/></param>
+        /// <param name="descriptors">A list of descriptors to be written.</param>
+        /// <param name="bulkUpdateGraph">An instance of the <see cref="BulkUpdateGraph"/>.</param>
+        /// <param name="oDataWriter">An instance of the <see cref="ODataWriterWrapper"/>.</param>
+        internal void WriteResourceSet(IReadOnlyList<Descriptor> descriptors, Dictionary<Descriptor, List<LinkDescriptor>> linkDescriptors, BulkUpdateGraph bulkUpdateGraph, ODataWriterWrapper oDataWriter)
+        {
+            ODataResourceSet resourceSet = new ODataResourceSet();
+            oDataWriter.WriteStart(resourceSet);
+
+            for (int i = 0; i < descriptors.Count; ++i)
+            {
+                Descriptor descriptor = descriptors[i];
+
+                if (descriptor is EntityDescriptor entityDescriptor)
+                {
+                    if (entityDescriptor.State == EntityStates.Deleted || entityDescriptor.State == EntityStates.Modified)
+                    {
+                        // We don't support Delete or Update in Deep insert
+                        // Throw exception?
+                    }
+                    else
+                    {
+                        this.WriteDeepInsertEntry(entityDescriptor, linkDescriptors, bulkUpdateGraph, oDataWriter);
+                    }
+                }
+                else if (descriptor is LinkDescriptor linkDescriptor)
+                {
+                    if (linkDescriptor.State == EntityStates.Deleted || linkDescriptor.State == EntityStates.Modified)
+                    {
+                        /// We don't support Delete or Update in Deep insert
+                        // Throw exception?
+                    }
+                    else
+                    {
+                        WriteODataId(linkDescriptor, oDataWriter);
+                    }
+                }
+            }
+
+            oDataWriter.WriteEnd();
+        }
+
+        /// <summary>
+        /// Writes the feed entry
+        /// </summary>
+        /// <param name="descriptors">A list of descriptors to be written.</param>

```

Comment:
Ensure this method retains a doc comment for linkDescriptors param like it did before.